### ISSUE TITLE
SIGABRT on SWIG_RubyUnlinkObjects

### ISSUE THREAD

**Issue by Sigill** [swig/swig#1168]
Hi everyone,

I'm trying to make an update of the [taglib-ruby](https://github.com/robinst/taglib-ruby) gem, which use Swig to wrap the [libtag](https://github.com/taglib/taglib) library in Ruby, but I'm facing a SIGABRT I don't understand.

This gem has [a unit test](https://github.com/robinst/taglib-ruby/blob/640fe2a8cea0936a0120b502c6c4e6bcf468df87/test/id3v2_memory_test.rb#L40) that triggers the Ruby garbage collector, in order to make sure native data are not leaked upon destruction of the Ruby wrapper object.

The destruction of the Ruby object triggers the `free_taglib_mpeg_file()`, which SIGABRT upon calling `SWIG_RubyUnlinkObjects()` on the underlying native object ([here](https://github.com/robinst/taglib-ruby/blob/640fe2a8cea0936a0120b502c6c4e6bcf468df87/ext/taglib_mpeg/taglib_mpeg_wrap.cxx#L2201)).

After various tests, it appear that Swig is causing the crash. A bisection pointed me to the 0e725b5d9bd534964ae606852453df46d04037ee commit, where the `SWIG_RubyUnlinkObjects` method was transformed from:

```c
void SWIG_RubyUnlinkObjects(void* ptr) {
  VALUE object = SWIG_RubyInstanceFor(ptr);

  if (object != Qnil) {
    DATA_PTR(object) = 0;
  }
}
```

to

```c
void SWIG_RubyUnlinkObjects(void* ptr) {
  VALUE object = SWIG_RubyInstanceFor(ptr);

  if (object != Qnil) {
    if (TYPE(object) != T_DATA)
      abort();
    DATA_PTR(object) = 0;
  }
}
```

The call to `abort()` being the source of the crash.

Does anyone know the reason for the `if (TYPE(object) != T_DATA) abort();` code? I plugged GDB on the failing unit test and in both cases (before and after the above commit), `TYPE(object)` is 30 (`T_DATA` being 12 on Ruby 2.3), so nothing appears to have changed (I'm fairly new to Swig and Ruby internals, so don't hesitate to tell me if I'm wrong).

**Comment by ojwb** [comment]
I'm not familiar with Ruby's C API, but I'd guess that `DATA_PTR()` is only valid for an object of type `T_DATA`, so this check is to ensure it's not called on other types of object.

But looking at the headers, object type 30 doesn't seem to be used by Ruby 2.3 (in Ruby 2.2 and earlier 30 was `RUBY_T_ZOMBIE`).  Are you definitely using Ruby 2.3?

**Comment by Sigill** [comment]
Hum, you have a good point.

On my Debian 9, I had older rubies remaining from system upgrades. The 2.3 version was the main one, but all the ruby tools (well, `rake`), were installed when the 2.1 release was the main one, and theirs shebangs were pointing to the `ruby2.1` binary.

I removed all old ruby packages & their associated dev packages, rebuilt Swig, regenerated & rebuilt the Swig wrappers, and now... `TYPE(object)` is 29 (`RUBY_T_ZOMBIE`). The mystery 30 is now gone, but the issue still here.

`DATA_PTR()` seems to be indeed only valid on `T_DATA` objects, but before the 0e725b5d9bd534964ae606852453df46d04037ee commit, non-data objects were simply ignored. @kkaempf, @wsfulton, you were referenced on this patch. Do you have some insights about the need for the `abort()`.

**Comment by ojwb** [comment]
Looks to me like before that commit, the object was just assumed to be `T_DATA` and its `DATA_PTR` set to zero regardless, presumably setting a somewhat arbitrary part of a different structure to zero.  We might have got away with that, but it's not a good thing to be doing.

Going only by the "ZOMBIE" part of the name, my guess is that the case you're hitting is something like a ruby object which has already been destroyed, and the code here should check for and skip `T_ZOMBIE`.  It'd be good if someone who actually knows this stuff could review this guess though.

**Comment by Sigill** [comment]
Indeed, the "non-data objects are simply ignored" solution is not the solution that was in place in the past, but the solution that seems to make sense now. I'll make a pull request.

**Comment by Sigill** [comment]
Actually, the error might be on the gem I'm playing with, because the [user provided `free` function](https://github.com/robinst/taglib-ruby/blob/640fe2a8cea0936a0120b502c6c4e6bcf468df87/ext/taglib_mpeg/taglib_mpeg_wrap.cxx#L2201) calls `SWIG_RubyUnlinkObjects()` on the object being destroyed.
According [this example](https://github.com/swig/swig/blob/069ce1f6e9c9238b7ca6667d8272245a65b5c701/Examples/ruby/free_function/example.i), and to some automatically generated free function, it does not seems to be required.

**Comment by Sigill** [comment]
After further investigations, I've been able to reproduce the issue using [this example](https://github.com/swig/swig/blob/069ce1f6e9c9238b7ca6667d8272245a65b5c701/Examples/ruby/free_function/example.i). This code will trigger the `abort()` (run it with `ruby -I. runme.rb`):

```ruby
require 'example'
t = Example::Animal.new("tiger1")
z = Example::Zoo.new
z.add_animal(t)
t = nil
GC.start
t = z.get_animal(0)
```

Turns out that when the Ruby GC is executed, the destruction of the GC'ed `T_DATA` objects is deferred: their type flag is first turned to `T_ZOMBIE`, then their free function is called (`SWIG_RubyUnlinkObjects()` in our case). `SWIG_RubyUnlinkObjects()` can therefore be called on non `T_DATA` objects, causing the issue.

Removing the `if (TYPE(object) != T_DATA) abort();` part introduced by commit 0e725b5d9bd534964ae606852453df46d04037ee appear to solve the issue. I don't think that setting the `DATA_PTR` of the ruby object to zero regardless of it's type is an issue: `SWIG_RubyUnlinkObjects()` can only be the free function of a `T_DATA` object. It might appear as a `T_ZOMBIE`, but that's just a flag, the object still has the structure of a `T_DATA` one.
